src/rulesets/Base/base.jl

```diff
@@ -3,15 +3,80 @@
 
 @scalar_rule one(x) zero(x)
 @scalar_rule zero(x) zero(x)
-@scalar_rule adjoint(x::Real) One()
 @scalar_rule transpose(x) One()
+
+# `adjoint`
+
+frule((_, Δz), ::typeof(adjoint), z::Number) = (z', Δz')
+
+function rrule(::typeof(adjoint), z::Number)
+    adjoint_pullback(ΔΩ) = (NO_FIELDS, ΔΩ')
+    return (z', adjoint_pullback)
+end
+
+# `real`
+
+@scalar_rule real(x::Real) One()
+
+frule((_, Δz), ::typeof(real), z::Number) = (real(z), real(Δz))
+
+function rrule(::typeof(real), z::Number)
+    # add zero(z) to embed the real number in the same number type as z
+    real_pullback(ΔΩ) = (NO_FIELDS, real(ΔΩ) + zero(z))
+    return (real(z), real_pullback)
+end
+
+# `imag`
+
 @scalar_rule imag(x::Real) Zero()
+
+frule((_, Δz), ::typeof(imag), z::Number) = (imag(z), imag(Δz))
+
+function rrule(::typeof(imag), z::Complex)
+    imag_pullback(ΔΩ) = (NO_FIELDS, real(ΔΩ) * im)
```

Comment:
I think we could generalize this to `z::Number` with this
```suggestion
function rrule(::typeof(imag), z::Number)
    # multiply by `one(z)` so for non-complex z, cotangent is similar to z
    imag_pullback(ΔΩ) = (NO_FIELDS, (real(ΔΩ) * im) * one(z))
```

Reply:
I agree that this should be generalised to `Real`, it might be best to implement a separate method as `imag(x::Real)` always returns `zero(x)`, hence the `pullback` should also just return `zero(x)`.

Reply:
There already is a `@scalar_rule` implementation for `imag`. My idea here is about generalizing beyond `Complex` to `Number`. But I don't think this is useful. How often does one define a non-real, non-complex type that is a valid input to `imag`? There isn't even an `imag(::Number)` method.

Reply:
Oh I see. I wouldn't worry about this -- sticking with `Complex` type constraint is fine.